### PR TITLE
Project 15

### PR DESCRIPTION
**Pull request by volodymyrtaras**

### CODE REVIEW

File: src/main/java/fifteen/entity/Project.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+            inverseJoinColumns = @JoinColumn(name = "company_id", referencedColumnName = "id")
+    )
+    private Set<Company> companies = new HashSet<>();
+
+    @ManyToMany
+    @JoinTable(name = "project_customer",
+            joinColumns = @JoinColumn(name = "project_id", referencedColumnName = "id"),
+            inverseJoinColumns = @JoinColumn(name = "customer_id", referencedColumnName = "id")
+    )
+    private Set<Customer> customers = new HashSet<>();
```
Comment (Anderwerty) [inline src/main/java/fifteen/entity/Project.java]: The same about Customer entity.

File: src/main/java/fifteen/entity/Project.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+            inverseJoinColumns = @JoinColumn(name = "developer_id", referencedColumnName = "id")
+    )
+    private Set<Developer> developers = new HashSet<>();
+
+    @ManyToMany
+    @JoinTable(name = "project_company",
+            joinColumns = @JoinColumn(name = "project_id", referencedColumnName = "id"),
+            inverseJoinColumns = @JoinColumn(name = "company_id", referencedColumnName = "id")
+    )
+    private Set<Company> companies = new HashSet<>();
```
Comment (Anderwerty) [inline src/main/java/fifteen/entity/Project.java]: The same about Company entity

File: src/main/java/fifteen/entity/Skill.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+import java.util.Set;
+
+@Slf4j
+@Getter
+@Setter
+@AllArgsConstructor
+@NoArgsConstructor
+@Entity
+@Table(name = "skills")
+public class Skill {
```
Comment (Anderwerty) [inline src/main/java/fifteen/entity/Skill.java]: Why not extended from entity Model?

File: src/main/java/fifteen/entity/Project.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
+package fifteen.entity;
+
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+import lombok.extern.slf4j.Slf4j;
+
+import javax.persistence.*;
```
Comment (Anderwerty) [inline src/main/java/fifteen/entity/Project.java]: Please don't use a wild card in imports. The same for other classes

File: src/main/java/fifteen/entity/Skill.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
+package fifteen.entity;
+
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+import lombok.extern.slf4j.Slf4j;
+
+import javax.persistence.*;
```
Comment (Anderwerty) [inline src/main/java/fifteen/entity/Skill.java]: The same

File: src/main/java/fifteen/entity/Company.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
+package fifteen.entity;
+
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+import lombok.extern.slf4j.Slf4j;
+
+import javax.persistence.*;
```
Comment (Anderwerty) [inline src/main/java/fifteen/entity/Company.java]: Please don't use a wild card in imports.

File: src/main/java/fifteen/entity/Developer.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
+package fifteen.entity;
+
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+import lombok.extern.slf4j.Slf4j;
+
+import javax.persistence.*;
```
Comment (Anderwerty) [inline src/main/java/fifteen/entity/Developer.java]: The same

File: src/main/java/fifteen/entity/Developer.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+@Setter
+@AllArgsConstructor
+@NoArgsConstructor
+@Entity
+@Table(name = "developers")
+public class Developer extends Model {
+    @Column(name = "age")
+    private Integer age;
+    @Column(name = "gender")
+    private String gender;
```
Comment (Anderwerty) [inline src/main/java/fifteen/entity/Developer.java]: In such case, you can represent male like "m", "M", "male",...

File: src/main/java/fifteen/entity/Developer.java [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    @Column(name = "salary")
+    private Double salary;
+
+    @ManyToMany
+    @JoinTable(name = "developer_skill",
+            joinColumns = @JoinColumn(name = "developer_id", referencedColumnName = "id"),
+            inverseJoinColumns = @JoinColumn(name = "skill_id", referencedColumnName = "id")
+    )
+    private Set<Skill> skills = new HashSet<>();
```
Comment (Anderwerty) [inline src/main/java/fifteen/entity/Developer.java]: Equals and hashCode?

File: src/main/java/fifteen/util/Hibernate.java [hunk 10]
```diff
@@ -0,0 +1,22 @@
+package fifteen.util;
+
+import javax.persistence.EntityManagerFactory;
+import javax.persistence.Persistence;
+
+public class Hibernate {
```
Comment (Anderwerty) [inline src/main/java/fifteen/util/Hibernate.java]: util class should be final and has a private constructor.
If you use Lombok you can use annotation UtilityClass
https://projectlombok.org/features/experimental/UtilityClass

File: src/main/java/fifteen/util/Hibernate.java [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+    private static final String PERSISTENCE_UNIT_NAME = "PERSISTENCE";
+    private static EntityManagerFactory factory;
+
+    public static EntityManagerFactory getEntityManagerFactory() {
```
Comment (Anderwerty) [inline src/main/java/fifteen/util/Hibernate.java]: Minor: Explain why it is not treading safely?

Comment (volodymyrtaras) [reply src/main/java/fifteen/util/Hibernate.java]: I do not know right now, but I am going to read about it

File: src/main/java/fifteen/entity/Project.java [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+@NoArgsConstructor
+@Entity
+@Table(name = "projects")
+public class Project extends Model {
+
+    @Column(name = "cost")
+    private Double cost;
+
+    @Column(name = "date")
+    private Date date;
```
[the remark this answers is not in the document]
Comment (volodymyrtaras) [reply src/main/java/fifteen/entity/Project.java]: To be honest, I did`nt understand, how to use it

File: src/main/java/fifteen/dao/CompanyDao.java [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    public boolean insertCompany(Company company) {
+        try {
+            entityManager.getTransaction().begin();
+            entityManager.persist(company);
+            entityManager.getTransaction().commit();
+            return true;
+        } catch (RuntimeException e) {
+            if (entityManager != null) {
```
[the remark this answers is not in the document]
Comment (volodymyrtaras) [reply src/main/java/fifteen/dao/CompanyDao.java]: @Anderwerty what to you mean?

**Comment by volodymyrtaras** [comment]
@Anderwerty fixed.

File: src/main/java/fifteen/dao/CompanyDaoImpl.java [hunk 14]
```diff
@@ -0,0 +1,75 @@
+package fifteen.dao;
+
+import fifteen.entity.Company;
+import lombok.extern.slf4j.Slf4j;
+
+import javax.persistence.EntityManager;
+
+@Slf4j
+public class CompanyDaoImpl implements CompanyDao {
```
Comment (Anderwerty) [inline src/main/java/fifteen/dao/CompanyDaoImpl.java]: image
This part of code looks like the same for all methods. Could you please move it into a seperate method and reuse.

File: src/main/java/fifteen/dao/DeveloperDaoImpl.java [hunk 15]
```diff
@@ -0,0 +1,76 @@
+package fifteen.dao;
+
+import fifteen.entity.Developer;
+import lombok.extern.slf4j.Slf4j;
+
+import javax.persistence.EntityManager;
+
+@Slf4j
+public class DeveloperDaoImpl implements DeveloperDao {
```
Comment (Anderwerty) [inline src/main/java/fifteen/dao/DeveloperDaoImpl.java]: The same

File: src/main/java/fifteen/dao/ProjectDaoImpl.java [hunk 16]
```diff
@@ -0,0 +1,75 @@
+package fifteen.dao;
+
+import fifteen.entity.Project;
+import lombok.extern.slf4j.Slf4j;
+
+import javax.persistence.EntityManager;
+
+@Slf4j
+public class ProjectDaoImpl implements ProjectDao {
```
Comment (Anderwerty) [inline src/main/java/fifteen/dao/ProjectDaoImpl.java]: The same

File: src/main/java/fifteen/entity/Project.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (Anderwerty) [reply src/main/java/fifteen/entity/Project.java]: Please read and google more about it.

**Comment by volodymyrtaras** [comment]
@Anderwerty fixed.

**Pull request merged**